geode-protobuf/src/test/java/org/apache/geode/protocol/RoundTripCacheConnectionJUnitTest.java

```diff
@@ -290,6 +291,59 @@ public void testConnectionCountIsProperlyDecremented() throws Exception {
   }
 
   @Test
+  public void testNewProtocolRespectsMaxConnectionLimit() throws IOException, InterruptedException {
+    cache.close();
+
+    CacheFactory cacheFactory = new CacheFactory();
+    Cache cache = cacheFactory.create();
+
+    CacheServer cacheServer = cache.addCacheServer();
+    final int cacheServerPort = AvailablePortHelper.getRandomAvailableTCPPort();
+    cacheServer.setPort(cacheServerPort);
+    cacheServer.setMaxConnections(16);
+    cacheServer.setMaxThreads(16);
+    cacheServer.start();
+
+    AcceptorImpl acceptor = ((CacheServerImpl) cacheServer).getAcceptor();
+
+    Socket[] sockets = new Socket[16];
+
+    for (int i = 0; i < 16; i++) {
+      Socket socket = new Socket("localhost", cacheServerPort);
+      sockets[i] = socket;
+      Awaitility.await().atMost(5, TimeUnit.SECONDS).until(socket::isConnected);
+      socket.getOutputStream()
+          .write(CommunicationMode.ProtobufClientServerProtocol.getModeNumber());
+    }
+
+    try (Socket socket = new Socket("localhost", cacheServerPort)) {
+      Awaitility.await().atMost(5, TimeUnit.SECONDS).until(socket::isConnected);
+      socket.getOutputStream()
+          .write(CommunicationMode.ProtobufClientServerProtocol.getModeNumber());
+      assertEquals(-1, socket.getInputStream().read());
+    }
+
+    for (Socket currentSocket : sockets) {
+      currentSocket.close();
+    }
+
+    Awaitility.await().atMost(5, TimeUnit.SECONDS)
```

Comment:
Could we inject some comments that explain on a higher level what we are trying to test?